---
title: "Vignette_2: Data Visualization with TCGA KIRP human study: LATE vs EARLY stage"
output:
  html_document:
    theme: united
    toc: yes
vignette: "%\\VignetteIndexEntry{Vignette} \n%\\VignetteEngine{knitr::rmarkdown} \n%\\VignetteEncoding{UTF-8}\n"
editor_options: 
  markdown: 
    wrap: 72
---

```{=html}
<style type="text/css">
.main-container {
  max-width: 1800px;
  margin-left: auto;
  margin-right: auto;
}
body{ /* Normal  */
      font-size: 18px;
  }
code.r{ /* Code block */
    font-size: 14px;
}
</style>
```

```{r, include = FALSE}
#https://bookdown.org/yihui/rmarkdown/html-document.html
#https://stackoverflow.com/questions/30446905/rmarkdown-font-size-and-header
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  knitr::opts_chunk$set(warning = FALSE, message = TRUE)
)
```
# StellarPath
Welcome to the second vignette of our software StellarPath. 
This vignette focuses on visualizing the data produced by StellarPath.
We hope to provide some ideas on how to investigate and elaborate our method's results.

## Note
Before checking this vignette, we suggest to go through the first vignette describing the workflow.
In this vignette, we will skip the explanations about operations and focus on the data interpretation and visualization.
In certain cases, we will limit the data to create figures that fit in the html vignette.
In a real application, we suggest to plot all the data in an local pdf.

## Case study
For this vignette, StellarPath is applied on a simple experimental study:

* We have: Kidney renal papillary cell carcinoma patient's samples divded by cancer stage: LATE vs EARLY
* We have: bulk RNA sequencing, small RNA sequencing and somatic mutation data for all the samples in study
* We want: deregulated pathways and molecules between LATE and EARLY cancer stage
* We want: information to further prioritize and investigate pathways and molecules
* We want: find a predictive pathway for EARLY cancer stage patients and check up the patient's similarities
* We want: find molecular markers for EARLY cancer stage patients
* We want: check up the pathway activity of all the patients in the significant pathway specific PSNs to detect subclasses

# Enviroment set up
Let us clean the R enviroment, set the working directory, load the software package, set the random seed generator for reproducing always the same results and define the number of cores available in the computer to run the operations.

* Be careful: set up the number of cores based on your resources, if you are not secure how to, then just set equal to 2
* Be careful: this vignette requires extra packages for plotting data


```{r setup}

#Clean workspace and memory ----
rm(list=ls())
gc()

#Set working directory ----
gps0=getwd()
gps0=paste(gps0,"/%s",sep="")
rootDir=gps0
setwd(gsub("%s","",rootDir))
set.seed(8)

#Load libraries ----
library("StellarPath")
SP_initialize()
library("qs")
library("data.table")
library("ggplot2")
library("pheatmap")
library("ggrepel")

#Set variables ----
#Input
#Set the number of cores available in the system
n_cores=15
#output
#Set the output file where all the data and results will be stored in the end
out_res_path="classif_KIRP.qs"
```

# Data set up
StellarPath is a binary classifier and works with one or multiple omics from high-throughput sequencing describing the two patient classes in comparison. For this vignette, we load the TCGA data that StellarPath includes in the package and extract the information about KIRP patients. Then, we run the main workflow on KIRP data.

```{r workflow}
#Load and set up the data
data("tcga_p2_l")
tcga_l=tcga_p2_l

#Set application-dependent parameters
name_dataset="KIRP" #Set name of the project
tax_id=9606 #specie id of the samples
groups=c("LATE","EARLY") #the name of the classes to compare (case vs control)

#Start monitor RAM memory
log_path=start_monitor_ram(name_dataset)
#Start monitor running time
startTime=Sys.time()

#Feed the sequencing raw data to prepare_data function that checks and prepare the user's input
data_l=prepare_data(info = tcga_l[[name_dataset]]$info[,c("patientID","pathologic_stage_binary")],
                    groups_name = groups,
                    gex = tcga_l[[name_dataset]]$omics$RNASeq2Gene,
                    miRNA = tcga_l[[name_dataset]]$omics$miRNASeqGene,
                    mutations = tcga_l[[name_dataset]]$omics$Mutation,
                    tax_id = tax_id,
                    n_cores = n_cores);tmp=gc();

#Find the pathways that in the training data produce a significant PSN
data_l=analyse_training(data_l)

#Classify the testing patients based on the significant pathway-based PSNs learnt from the training
data_l=classify_testing(data_l)

#End monitor resources
runningTime=convert_min2h(difftime(Sys.time(), startTime, units='mins')[[1]])
max_ram=get_max_ram_used(log_path)
resources_l=list(runningTime=runningTime, max_ram=max_ram)
#Update the performance list
data_l$performances$performances$name_dataset=name_dataset
data_l[["performances"]]=c(data_l[["performances"]],resources_l)

#Report the ending of the classification
cat("Classification performed in", runningTime, "hours and with", max_ram, "GB of RAM")

#Retrieve best significant pathway-based PSNs with the predicted patients and the training results (best predictive markers)
data_l=enrichment_analysis(data_l)

#Save results
qsave(data_l,file = out_res_path, nthreads = n_cores)

```

# Data Extraction
Let us extract the data and results produced by StellarPath

```{r data extraction}
#Extract differentially expressed and stable molecules (e.g. genes)
significant_genes_df=data_l$enrichment$omics$gex$molecule_analysis

#Extract and format the classification performance of the GCN with each significant pathway specific PSN
perfsXpathway_dfs=data_l$performances$perfsXpathway_dfs

#Extract significant pathway specific PSN
enriched_pathway_df=unique(data_l$enrichment$pathway_analysis$top_pathways_df[,seq(1,13)])

#Extract how much each patient is central in the pathway specific PSNs
#Here we will extract the centrality of every patient in every significant pathway
cts_l=list()
for(i_path in 1:nrow(enriched_pathway_df)){
    pathway_data_l=get_pathway_data(data_l,enriched_pathway_df$path2pathway[i_path])
    cts_l[[i_path]]=pathway_data_l$centrality
}
cts_df=as.data.frame(data.table::rbindlist(cts_l,use.names = TRUE))
rownames(cts_df)=enriched_pathway_df$pathways
```

# Plot
We can now start to produce the plots and visualize the data about the classification and the patients.

## GCN performance per PSN
We want to understand how much the GCN performed with each pathway specific PSN.

```{r performances, out.width = "100%", fig.align = "center"}
perfsXpathway_dfs$dataset=name_dataset
perfsXpathway_dfs$method="StellarPath"
perfsXpathway_dfs=perfsXpathway_dfs[!is.na(perfsXpathway_dfs$test_mcc),]
ggplot(perfsXpathway_dfs, aes(x=dataset, y=test_mcc, fill=method)) + 
  geom_violin(color="grey35",lwd=0.3,scale="width") +
  scale_fill_manual(values = c('#f6b45e')) +
  theme(text = element_text(size=7)) +
  labs(y = "Matthews correlation coefficient", x="TCGA datasets") +
  facet_grid(~ dataset, scales = "free") +
  theme(axis.ticks.x=element_blank(), axis.text.x=element_blank()) +
  ylim(0.5, 1)
```
The GCN managed to predict very good the testing patients in most of the significant pathway specific PSNs involved in the different runs of cross-validation. This means that we can consider the results coming out this classification because pathways and molecules will be both deregulated and predictive.

## Pathway Ranking
We plot the statistics of the significant pathway specific PSNs.
This plot allows to summarize the most important statistics about PSNs and let us prioritize which pathways to study further.
Precisely, we plot:

* Separability Power: (how much the classes in study are separated in the PSN and how much one class is cohesive while the second one is sparse)
* Group: the patient's class that is cohesive in the PSN and also the class whose molecular markers are deregulated in the pathway
* Regulation Type: if the pathway is activated or inhibited by the deregulated molecules of the cohesive class
* Group's cohesion: indicates the smallest similarity between the patients of the cohesive class

```{r ranking, results='asis', out.width = "100%", fig.align = "center", message=FALSE}
x=enriched_pathway_df
x=x[order(x$power),]
x=x[1:50,]
x$pathways=factor(x$pathways,ordered = TRUE,levels = x$pathways)
x=na.omit(x)
ggplot(x, aes(y=pathways, x=power, color=regulation, shape=group, size=min_intra_STRONG)) +
  geom_point() + theme_bw() + 
  scale_color_manual(breaks = c("activated", "inhibited"), values=c("red", "blue")) +
  labs(x="PSN's Separability Power", y="Pathway specific PSNs", 
       color="Regulation Type", size="PSN's Class\ncohesion") +
  theme(text = element_text(size=7)) +
  scale_size(range = c(0, 2))
```

## PSN graph plot
The TNFα signaling via NFκB pathway is involved in late cancer stage by promoting cancer cell migration and invasion1. The NFκB pathway is a key signaling pathway that regulates the expression of genes involved in inflammation, immunity, cellular homeostasis and tumour progression1. Excessive activation of the NFκB-signaling pathway has been documented in various tumor tissues2. KIAA1522 hyperactivates TNFα-NFκB signaling to facilitate resistance to platinum reagents. Targeting NFκB signaling through small molecule inhibitors may be a rational strategy to conquer chemoresistance and synergize platinum-based chemotherapy in KIAA1522 overexpressed lung adenocarcinomas3.

```{r PSN, results='asis', out.width = "100%", fig.align = "center", message=FALSE}
#Watch the stats of the pathway of interest
path_name="HALLMARK_E2F_TARGETS"
knitr::kable(enriched_pathway_df[enriched_pathway_df$pathways==path_name,])

#Retrieve the data of the pathway and its PSN
data4path=get_pathway_data(data_l,enriched_pathway_df$path2pathway[enriched_pathway_df$pathways==path_name])
info=data_l$enrichment$info
pl_l=plot_PSN(info,data4path$PSN,k_clusters = 10,edge_threshold = 0.4,GLratio = 3)
```
The following LATE patients are the most important for their class:

* LATE.1
* LATE.7
* LATE.21
* LATE.23
* LATE.35
* LATE.39

While it seems that the patient LATE.26 is not regulating the pathway as the other LATE patients.
At this point, we want to remember that StellarPath function represents only few patients in the network to visualize.
This means that there can be other patients behind LATE.26, let's see them.

```{r PSN outliers, results='asis'}
pl_l$names_in_k_l$LATE.26
```

If we want to get the original names of the patient's samples which can be outliers for the pathways:

```{r PSN original names, results='asis'}
plyr::mapvalues(pl_l$names_in_k_l$LATE.26, from=info$IDs, to=rownames(info), warn_missing = FALSE)
```

## Molecular Markers
We plot the statistics of the significant molecules that StellarPath found deregulated in expression and stability in the CELL KILLING pathway.
This plot allows to summarize the most important statistics about the deregulated molecules and let us prioritize which molecules to study further.
Precisely, we plot:

* logFoldChange: how much the molecule is deregulated in term of expression values in the comparison LATE vs EARLY
* Stability: how much the molecule is deregulated in term of variance of the expression values in the comparison LATE vs EARLY (more the value is lower and more the molecule has a more stable expression in EARLY patients)
* Activity_EARLY: Relative expression value of the molecule in the EARLY class

```{r Markers, results='asis',out.width = "100%", fig.align = "center", message=FALSE}
#Extract the set of molecules that enriched the pathway of interest
mls=data4path$set
#Extract the stats about the molecules
sdr_mls=significant_genes_df[mls,]
sdr_mls$names=rownames(sdr_mls)

ggplot(sdr_mls, aes(y=logFoldChange, x=stability, color=activity_LATE)) +
  geom_point() + theme_bw() +
  labs(x="Stability", y="logFoldChange", 
       color="Relative expression\n in LATE class") +
  theme(text = element_text(size=7)) +
  geom_label_repel(aes(label = names),
                   box.padding   = 0, 
                   point.padding = 0.1,
                   size = 5, alpha=0.8, max.overlaps=5, label.size = NA, fill = "white",
                   segment.color = 'grey50')
```
The gene MSC seems to be the most important deregulated markers in the pathway for LATE cancer stage patients.

## Centralities in pathway specific PSNs
In the end, we also want to plot how much each patient is central in each pathway specific PSN.
This plot allows to understand if there are subclasses inside the classes in study or there are outliers.
For example, we have suclasses of LATE cancer stage patients if a large subset of patients is central in only some pathways and not central in others. We don't have subclasses if LATE cancer stage patients are all central in all the significant pathways. We have outlier patients if a small subset of LATE patients is not central in only some pathways.

```{r centralities, out.width = "160%", fig.align = "center"}
#We set up the annotation dataframe for the columns/samples in order to see their class in the plot
classes=sub("\\..*", "",colnames(cts_df))
anno=data.frame(ids=colnames(cts_df),classes=classes,samples="NA")
rownames(anno)=anno$ids
anno=anno[,-1]

#We plot a subset of the centrality matrix to simplify the plot for the vignette
pheatmap(
  cts_df[,sample(seq(1,ncol(cts_df)))[1:70]], scale = "row",
  main=paste("StellarPath patient's centralities:",name_dataset),
  fontsize=8,
  fontsize_row=3.5,
  fontsize_col=3.5,
  angle_col=45,
  annotation_col = anno
)
```

At the same time, we can see already that there are clearly two subclasses in both LATE and EARLY cancer stage patients.
Analysing the centrality matrix with clustering techniques could further allow us to be more specific, detect the subclasses and distinguish them based on the pathways.




